core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -725,10 +740,15 @@ private[spark] class ExecutorAllocationManager(
         if (stageIdToNumRunningTask.contains(stageId)) {
           stageIdToNumRunningTask(stageId) += 1
         }
-        // This guards against the race condition in which the `SparkListenerTaskStart`
-        // event is posted before the `SparkListenerBlockManagerAdded` event, which is
-        // possible because these events are posted in different threads. (see SPARK-4951)
-        if (!allocationManager.executorIds.contains(executorId)) {
+        // This guards against the following race condition:
+        // 1. The `SparkListenerTaskStart` event is posted before the
+        // `SparkListenerExecutorAdded` event
+        // 2. The `SparkListenerExecutorRemoved` event is posted before the
+        // `SparkListenerTaskStart` event
+        // Above cases are possible because these events are posted in different threads.
+        // (see SPARK-4951 SPARK-26927)
+        if (!allocationManager.executorIds.contains(executorId) &&
+          !allocationManager.removedExecutorIds.contains(executorId)) {
```

Comment:
It's possible that `removedExecutorIds` never get trimmed. So, when an executor, which killed due to idle and revived due to tasks backlogged, comes with the `SparkListenerTaskStart` event, it could not be handled correctly. Right?

Reply:
@Ngone51  

> when an executor, which killed due to idle and revived due to tasks backlogged
Why can a removed executor revive again? Driver should not recognize removed executors.
So we should not add it to the `allocationManager.executorIds`.

Reply:
oh, I see. We'll always have unique id for executor, right ?

Reply:
Yes, that's it.

Reply:
I wonder if you couldn't achieve the same here by asking the scheduler backend whether the executor is known? `SparkListenerExecutorRemoved` is posted after the executor is removed from the internal state, so as far as I can see that would be the same check without needing to keep the extra state here.

Reply:
I thought that my suggestion might affect the other race ("task start before executor added" race), but I think it would be fine; I don't believe the scheduler would post a task start if the executor was not tracked internally already.

Reply:
@vanzin 

> I don't believe the scheduler would post a task start if the executor was not tracked internally already.

It can happen, because now the `ListenerBus` works with several `AsyncEventQueue`, the `allocationManager.executorIds` differs from `CoarseGrainedSchedulerBackend.exectuorDataMap` in that `allocationManager.executorIds` is added when `SparkListenerExecutorAdded` polled and processed in `AsyncEventQueue`.
So executor allocation is actually not kept sync with task scheduling for now.

Reply:
Or we change the `AsyncEventQueue` for executor management to a SyncQueue  which directly calls listener's callbacks.
e.g:
```
class SyncQueue extends SparkListenerBus {

def post(event: SparkListenerEvent): Unit = {
    super.postToAll(event)
    ....
}

}
```

Reply:
You didn't understand my suggestion.

I'm asking you to, instead of keeping a list of executors that have been removed here, just as the scheduler about whether that executor exist. The scheduler keeps the authoritative lists of executors.

Reply:
OK, I got it and I think it's a good solution. I have updated it and fix some tests, thank you so much! @vanzin